components/style/parallel.rs

```diff
@@ -71,57 +71,64 @@ pub fn traverse_dom<N, D>(traversal: &D,
 #[allow(unsafe_code)]
 fn top_down_dom<'a, 'scope, N, D>(unsafe_nodes: &'a [UnsafeNode],
                                   root: OpaqueNode,
-                                  mut data: PerLevelTraversalData,
+                                  mut traversal_data: PerLevelTraversalData,
                                   scope: &'a rayon::Scope<'scope>,
-                                  traversal: &'scope D)
+                                  traversal: &'scope D,
+                                  tls: &'scope rayon::ScopedTLS<'scope, D::ThreadLocalContext>)
     where N: TNode,
           D: DomTraversal<N>,
 {
     let mut discovered_child_nodes = vec![];
-    for unsafe_node in unsafe_nodes {
-        // Get a real layout node.
-        let node = unsafe { N::from_unsafe(&unsafe_node) };
-
-        // Perform the appropriate traversal.
-        let mut children_to_process = 0isize;
-        traversal.process_preorder(node, &mut data);
-        if let Some(el) = node.as_element() {
-            D::traverse_children(el, |kid| {
-                children_to_process += 1;
-                discovered_child_nodes.push(kid.to_unsafe())
-            });
+    {
+        // Scope the borrow of the TLS so that the borrow is dropped before
+        // potentially traversing a child on this thread.
+        let mut tls_borrow = tls.borrow_mut();
+        if tls_borrow.is_none() {
```

Comment:
I added a .ensure() function. I actually wanted to do so originally, but couldn't because we don't have AtomicRefMut::map. Now that we're using regular RefCell, I can use RefMut::map.